Support PHP 8.1 & PHP 8.2

craig410 opened the pull request.

craig410 committed:
Introduce support and test on PHP 8.1

Dropping PHP 7.4

craig410 committed:
Configure PHPUnit to convert deprecation notices to exceptions

craig410 committed:
Classes implementing ArrayAccess must adhere to strict return types

craig410 committed:
PHPUnit assertRegExp is deprecated in favour of assertMatchesRegularExpression

craig410 committed:
Replace usages of strftime

craig410 committed:
Fix deprecation warnings where PHP functions are being provided with incorrect variable types

craig410 commented:
@acoulton it's taking shape, smaller changeset than I imagined. WDYT? ~I know it's failing on the lowest dependency builds. Haven't looked at it properly yet, appears to be some kohana autoloader related thing not finding classes~ PHPUnit 9.5 below 9.5.22 breaks loading PSR-0 classes

craig410 committed:
Bump phpunit to 9.5.22 as versions below break loading of PSR-0/PEAR style classes

craig410 committed:
Update changelog

craig410 committed:
Test and support PHP 8.2

acoulton commented on classes/Kohana/Arr.php:
```diff
@@ -127,7 +127,7 @@ public static function path($array, $path, $default = NULL, $delimiter = NULL)
+			if (\ctype_digit((string) $key))
 			{
 				// Make the key an integer
 				$key = (int) $key;
```
I'm not actually convinced this serves any purpose, afaik PHP `$array['1']` and `$array[1]` refer to the same thing, so maybe we could safely drop the whole if block/cast.

craig410 replied:
```php
$arr = [1 => 'foo', '2' => 'bar'];

echo $arr[1];
echo $arr[2];

echo $arr['1'];
echo $arr['2'];
``` 

results in 

`foobarfoobar`

I would have at least expected `echo $arr[2]` to fail. We'll have strictly typed arrays appear in 8.something

acoulton replied:
Nah that seems how I'd expect it to be. That will be a killer of a deprecation/BC break if they change that!

Tbf we're lucky that `$arr['foo']` doesn't resolve to `$arr[0]` if the array happens to be an indexed one :)

craig410 replied:
I only realised the other day in something else you could have indexed and associative keys in the same array. The return array of getimagesize is something like `[0=>1920, 1=>1080, 'mime'=>'jpeg']` where 0 is width, 1 is height and mime does what it says on the tin. I guess it's to facilitate list() assuming you just wanted the width and height

acoulton replied:
Yeah, it's kinda nuts. Things like PDO::FETCH_BOTH and preg_match with named captures do similar things ISTR.

acoulton commented on classes/Kohana/Valid.php:
```diff
@@ -346,13 +354,17 @@ public static function luhn($number)
 	 */
 	public static function phone($number, $lengths = NULL)
 	{
+		if ( ! is_string($number)){
```
Worth the changelog mentioning that all the Valid:: methods for numeric strings (phone/card/alpha/digit etc) no longer validate int / float / values even if they would be valid in string form?

acoulton replied:
**actually** I've just had a thought, although kohana validation is mostly used for POST, if it's used for e.g. entity validation (which ISTR we still do some places) then it may get integer etc input to the numeric validators.

So probably we do actually want things like `Valid::decimal`, `Valid::digit` at least to do something more like:
```
if (is_int($number) || is_float($number)) { 
  $number = string($number);
}
```

Which is a bit backwards really but will avoid risk of this breaking non-POST validators.

Sorry, realise that's the opposite of what I said yesterday.

craig410 replied:
see 4c50074

acoulton replied:
Ta

acoulton reviewed (commented):
This is looking great.

craig410 committed:
Numeric validators continue to accept integers / floats and treat as strings

craig410 committed:
Remove redundant type cast

craig410 commented:
@acoulton Release this as 4.8.7 or 4.9.0?

craig410 committed:
Require mbstring extension

craig410 committed:
Do not provide fallback for mb_string functions

craig410 committed:
Exclude guide and media from releases

craig410 committed:
Fix dynamic property assignment warning by not setting it

If the only place Request::_routes is ever set is from here - and it's never read... this must be left behind from previous refactorings

craig410 committed:
Fix typo causing dynamic assignment warning

craig410 committed:
Just a hunch - implode wants a separator

craig410 committed:
Fix strcasecomp to match underlying stdlib strcasecomp

craig410 committed:
Update changelog

craig410 committed:
Merge pull request #42 from ingenerator/experimental-test-php82

Experimental test on PHP 8.2
